test/bundler/karma-stencil-utils.ts

```diff
@@ -0,0 +1,219 @@
+          const tempScripts: NodeListOf<HTMLScriptElement> = testElement.querySelectorAll('script');
+          for (let i = 0; i < tempScripts.length; i++) {
+            const script: HTMLScriptElement = document.createElement('script');
+            if (tempScripts[i].src) {
+              script.src = tempScripts[i].src;
+            }
+            if (tempScripts[i].hasAttribute('nomodule')) {
+              script.setAttribute('nomodule', '');
+            }
+            if (tempScripts[i].hasAttribute('type')) {
+              const typeAttribute = tempScripts[i].getAttribute('type');
+              if (typeof typeAttribute === 'string') {
+                // older DOM implementations would return an empty string to designate `null`
+                // here, we interpret the empty string to be a valid value
+                script.setAttribute('type', typeAttribute);
+              }
+            }
+            script.innerHTML = tempScripts[i].innerHTML;
+
+            if (tempScripts[i].parentNode) {
+              // the scripts were found by querying a common parent node, which _should_ still exist
+              tempScripts[i].parentNode!.insertBefore(script, tempScripts[i]);
+              tempScripts[i].parentNode!.removeChild(tempScripts[i]);
+            } else {
+              // if for some reason the parent node no longer exists, something's manipulated it while we were parsing
+              // the script tags. this can lead to undesirable & hard to debug behavior, fail.
+              reject('the parent node for script tags no longer exists. exiting.');
+            }
+          }
+        };
+
+        parseAndRebuildScriptTags();
+
+        /**
+         * Create a listener for Stencil's "appload" event to signal to the test framework the application and its
+         * children have finished loading
+         */
+        const onAppLoad = () => {
+          window.removeEventListener('appload', onAppLoad);
+          allReady().then(() => {
+            resolve(testElement);
+          });
+        };
+        window.addEventListener('appload', onAppLoad);
+      };
+
+      /**
+       * Ensure that all `onComponentReady` functions on Stencil elements in the DOM have been called before rendering
+       * @returns an array of promises, one for each `onComponentReady` found on a Stencil component
+       */
+      const allReady = (): Promise<d.HTMLStencilElement[] | void> => {
+        const promises: Promise<d.HTMLStencilElement>[] = [];
+
+        /**
+         * Function that recursively traverses the DOM, looking for Stencil components. Any `componentOnReady`
+         * functions found on Stencil components are pushed to a buffer to be run after traversing the entire DOM.
+         * @param promises the buffer of promises to add instances of `componentOnReady` to
+         * @param elm the current element being inspected
+         */
+        const waitForDidLoad = (promises: Promise<any>[], elm: Element): void => {
```

Comment:
should `promises` have type `Promise<d.HTMLStencilElement>[]` here? also, do we need to pass it in to the function, since it should be in-scope for `waitForDidLoad` anyhow?

Reply:
> should promises have type Promise<d.HTMLStencilElement>[] here?

Yeah - good catch!

> also, do we need to pass it in to the function, since it should be in-scope for waitForDidLoad anyhow?

nope, I don't think so - I was originally going to change the scoping here, but thought better of it. Fixed in de7ec55f9